crypto/rsa_extra/rsassa_pss_asn1.c

```diff
@@ -0,0 +1,317 @@
+          (CBS_len(params) == 0));
+}
+
+// Decode One-way Hash Functions: return one on success and zero on failure.
+// See 2.1. https://tools.ietf.org/html/rfc4055#page-5
+static int decode_one_way_hash(CBS *cbs, RSA_ALGOR_IDENTIFIER **hash_algor) {
+  CBS seq, oid;
+  if (CBS_get_asn1(cbs, &seq, CBS_ASN1_SEQUENCE) && (CBS_len(cbs) == 0) &&
+      CBS_get_asn1(&seq, &oid, CBS_ASN1_OBJECT) && is_absent_or_null(&seq) &&
+      parse_oid(&oid, rsa_pss_hash_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_hash_functions), hash_algor)) {
+    return 1;
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  return 0;
+}
+
+// Decode Mask Generation Functions: return one on success and zero on failure.
+// See 2.2. https://tools.ietf.org/html/rfc4055#page-5
+static int decode_mask_gen(CBS *cbs, RSA_MGA_IDENTIFIER **mga) {
+  CBS seq, mgf1_oid, hash_seq, hash_oid;
+  RSA_ALGOR_IDENTIFIER *mgf1 = NULL;
+  RSA_ALGOR_IDENTIFIER *hash_algor = NULL;
+  if (CBS_get_asn1(cbs, &seq, CBS_ASN1_SEQUENCE) && (CBS_len(cbs) == 0) &&
+      CBS_get_asn1(&seq, &mgf1_oid, CBS_ASN1_OBJECT) &&
+      CBS_get_asn1(&seq, &hash_seq, CBS_ASN1_SEQUENCE) &&
+      (CBS_len(&seq) == 0) &&
+      CBS_get_asn1(&hash_seq, &hash_oid, CBS_ASN1_OBJECT) &&
+      is_absent_or_null(&hash_seq) &&
+      parse_oid(&mgf1_oid, rsa_pss_mg_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_mg_functions), &mgf1) &&
+      parse_oid(&hash_oid, rsa_pss_hash_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_hash_functions), &hash_algor)) {
+    *mga = RSA_MGA_IDENTIFIER_new();
+    if (*mga != NULL) {
+      (*mga)->mask_gen = mgf1;
+      (*mga)->one_way_hash = hash_algor;
+      return 1;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  RSA_ALGOR_IDENTIFIER_free(mgf1);
+  RSA_ALGOR_IDENTIFIER_free(hash_algor);
+  return 0;
+}
+
+// get_context_specific_value sets |*out| to the contents of DER-encoded, ASN.1
+// element (not including tag and length bytes) and advances |seq| over it.
+// When the tag value does not exist, |seq| gets recovered.
+// It returns one when the element exists.
+static int get_context_specific_value(CBS *seq, CBS *out, int index) {
+  unsigned int tag_value = CBS_ASN1_CONTEXT_SPECIFIC | CBS_ASN1_CONSTRUCTED | index;
+  CBS seq_cp = {seq->data, seq->len};
+  if (CBS_get_asn1(seq, out, tag_value)) {
+    return 1;
+  } else {
+    // All fields of RSASSA-PSS-params can be absent, which means default.
+    // Recover |seq|.
+    seq->data = seq_cp.data;
+    seq->len = seq_cp.len;
+    return 0;
+  }
+}
+
+// Decode [0] HashAlgorithm of RSASSA-PSS-params: return one on success and zero
+// on failure. See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_hash(CBS *seq, RSA_ALGOR_IDENTIFIER **hash_algor) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 0)) {
+    // HashAlgorithm field can be absent, which means default.
+    return 1;
```

Comment:
The field being absent is valid case. The failure case will recover `CBS seq`, and let these bytes be parsed by other cases `[1], [2], [3]`. At the end of pss param decode, one of success criteria is to check if all bytes are successfully parsed --`CBS_len(seq) == 0`.

Reply:
`get_context_specific_value` can fail because the encoding is wrong. Do we actually want to assume that if this function fails, whatever we try to parse is absent?

I see, that the `CBS` API might not increment the "read" pointer if there is a parsing error, and we would therefore error later anyway. However, this is not clear when reading the code.

In addition, if there is a parsing error and the `seq` is recovered to its original state, then all subsequent parsing attempts would error.

If we can return as soon as we see an error, I think that would be the preferred behaviour.

Reply:
> get_context_specific_value can fail because the encoding is wrong. Do we actually want to assume that if this function fails, whatever we try to parse is absent?

This function returns `1` when the element exists, otherwise, return `0`. No success or failure.

`CBS_get_asn1(seq, out, tag_value)` returning `0` can confirm the `tag_value` not found, right?

> I see, that the CBS API might not increment the "read" pointer if there is a parsing error, and we would therefore error later anyway. However, this is not clear when reading the code.

* verified CBS API behavior: If pss params only have `[1]`, without this recover,  after checking `[0]`, the `CBS_len(seq) == 0`.

> If we can return as soon as we see an error, I think that would be the preferred behaviour.

This may require CBS API optimization/profiling.

Reply:
It seems wrong to me, that we continue parsing even though we have discovered a wrong encoding. I don't think we have reason about what can go wrong and what cannot. I'm taking a paranoid stand here.

We should at least look into what it takes to bail early - what changes in the code is required? What does other places in the code do when using `get_context_specific_value`?

Reply:
Tracked in CryptoAlg-677

Reply:
Can you add the bytes that include wrong encoding? 677 seems vague to me.